src/vm.rs

```diff
@@ -1480,6 +1496,56 @@ impl<'a, T: GuestStagePagingMode> FinalizedVm<'a, T> {
         Ok(0)
     }
 
+    fn handle_salus_test(
+        &self,
+        test_func: SalusTestFunction,
+        _active_pages: &ActiveVmPages<T>,
+    ) -> EcallResult<u64> {
+        use SalusTestFunction::*;
+        match test_func {
+            MemCopy(args) => self.guest_test_memcopy(args.to, args.from, args.len),
+        }
+    }
+
+    fn map_guest_range_in_umode_slot(
+        &self,
+        slot: u64,
+        addr: u64,
+        len: u64,
+        writable: bool,
+    ) -> EcallResult<(u64, GuestUmodeMapping)> {
+        let base = PageSize::Size4k.round_down(addr);
+        let end = addr
+            .checked_add(len)
+            .ok_or(EcallError::Sbi(SbiError::InvalidParam))?;
+        let umode_mapping = self
+            .vm_pages()
+            .map_in_umode_slot(
+                slot,
+                self.guest_addr_from_raw(base)?,
+                PageSize::num_4k_pages(end - base),
+                writable,
+            )
+            .map_err(EcallError::from)?;
+        let vaddr = umode_mapping.vaddr().bits() + (addr - base);
+        Ok((vaddr, umode_mapping))
+    }
+
+    fn guest_test_memcopy(&self, to: u64, from: u64, len: u64) -> EcallResult<u64> {
+        let overlap = core::cmp::max(to, from) <= core::cmp::min(to + len - 1, from + len - 1);
```

Comment:
Perhaps one option is to move the overlap checks to the unsafe function and enforce them there? This will be one less requirement on the caller, and it's an explicit check as well.

Reply:
Yep, good point. A bit hairy because i can't really use API error there. but did that.